### PR TITLE
Fix a bug where gRPC message deframer is closed before writing respon…

### LINKED ISSUE #2947: Kotlin gRPC service with blocking execution context: not returning result with multiple concurrent calls
When I execute multiple concurrent calls to a service, which is based on Kotlin coroutines and uses blockingExecutionContext, some calls return OK status but no response. In this case client reports an error (I tried both java & kotlin clients - error message is slightly different, but result is the same). With the non-blocking context there's no issue.

**Steps to reproduce:**

In armeria grpc-kotlin examples do the following changes:
1) in service function HelloServiceImpl.blockingHello decrease the sleep timeout value (e.g. make it Thread.sleep(10)). (Interesting, that all works OK with bigger timeouts (eg 500))
2) Change test method HelloServiceTest.replyFromServerSideBlockingCall to the following:
```
    @ParameterizedTest
    @MethodSource("uris")
    fun replyFromServerSideBlockingCall(uri: String) {
        runBlocking {
            val helloService = Clients.newClient(uri, HelloServiceCoroutineStub::class.java)
            repeat(30) {
                launch {
                    assertThat(helloService.blockingHello(HelloRequest.newBuilder().setName("Armeria").build()).message)
                            .isEqualTo("Hello, Armeria!")
                }
            }
        }
    }
```

After running this test, I get a consistent exception:

```
INTERNAL: Expected one request for MethodDescriptor{fullMethodName=example.grpc.hello.HelloService/BlockingHello, type=UNARY, idempotent=false, safe=false, sampledToLocalTracing=true, requestMarshaller=io.grpc.protobuf.lite.ProtoLiteUtils$MessageMarshaller@5298be71, responseMarshaller=io.grpc.protobuf.lite.ProtoLiteUtils$MessageMarshaller@29f896c1, schemaDescriptor=example.armeria.grpc.kotlin.HelloServiceGrpc$HelloServiceMethodDescriptorSupplier@72c26864} but received none
io.grpc.StatusException: INTERNAL: Expected one request for MethodDescriptor{fullMethodName=example.grpc.hello.HelloService/BlockingHello, type=UNARY, idempotent=false, safe=false, sampledToLocalTracing=true, requestMarshaller=io.grpc.protobuf.lite.ProtoLiteUtils$MessageMarshaller@5298be71, responseMarshaller=io.grpc.protobuf.lite.ProtoLiteUtils$MessageMarshaller@29f896c1, schemaDescriptor=example.armeria.grpc.kotlin.HelloServiceGrpc$HelloServiceMethodDescriptorSupplier@72c26864} but received none
	at io.grpc.kotlin.HelpersKt$singleOrStatusFlow$1.invokeSuspend(Helpers.kt:73)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.internal.ScopeCoroutine.afterResume(Scopes.kt:32)
	at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.kt:113)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
.....etc...
```

I am not sure if this is a problem in Armeria itself or in kotlin-grpc...

### PR DESCRIPTION
**Pull request by ikhoon**
…ses.

Motivation:

This is a workaround that some gRPC implementations do not handle `onHalfClose` properly.
For example, sporadically the listener of gRPC-Kotlin calls ArmeriaServerCall.close() due to some race condition of a Coroutine channel of the request stream.
```java
	at com.linecorp.armeria.server.grpc.ArmeriaServerCall.close(ArmeriaServerCall.java:286)
	at io.grpc.kotlin.ServerCalls$serverCallListener$1.invoke(ServerCalls.kt:257)
	at io.grpc.kotlin.ServerCalls$serverCallListener$1.invoke(ServerCalls.kt:46)
	at kotlinx.coroutines.InvokeOnCompletion.invoke(JobSupport.kt:1386)
	at kotlinx.coroutines.JobSupport.notifyCompletion(JobSupport.kt:1529)
	at kotlinx.coroutines.JobSupport.completeStateFinalization(JobSupport.kt:323)
	at kotlinx.coroutines.JobSupport.finalizeFinishingState(JobSupport.kt:240)
	at kotlinx.coroutines.JobSupport.tryMakeCompletingSlowPath(JobSupport.kt:903)
	at kotlinx.coroutines.JobSupport.tryMakeCompleting(JobSupport.kt:860)
	at kotlinx.coroutines.JobSupport.makeCompletingOnce$kotlinx_coroutines_core(JobSupport.kt:825)
	at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.kt:111)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
	at kotlinx.coroutines.DispatchedTaskKt.resume(DispatchedTask.kt:175)
	at kotlinx.coroutines.DispatchedTaskKt.resumeUnconfined(DispatchedTask.kt:137)
	at kotlinx.coroutines.DispatchedTaskKt.dispatch(DispatchedTask.kt:108)
	at kotlinx.coroutines.CancellableContinuationImpl.dispatchResume(CancellableContinuationImpl.kt:308)
	at kotlinx.coroutines.CancellableContinuationImpl.completeResume(CancellableContinuationImpl.kt:395)
	at kotlinx.coroutines.channels.AbstractChannel$ReceiveHasNext.resumeReceiveClosed(AbstractChannel.kt:913)
	at kotlinx.coroutines.channels.AbstractSendChannel.helpClose(AbstractChannel.kt:312)
	at kotlinx.coroutines.channels.AbstractSendChannel.close(AbstractChannel.kt:241)
	at kotlinx.coroutines.channels.SendChannel$DefaultImpls.close$default(Channel.kt:102)
	at io.grpc.kotlin.ServerCalls$serverCallListener$2.onHalfClose(ServerCalls.kt:288)
	at com.linecorp.armeria.server.grpc.ArmeriaServerCall.invokeHalfClose(ArmeriaServerCall.java:426)
	at com.linecorp.armeria.server.grpc.ArmeriaServerCall.endOfStream(ArmeriaServerCall.java:419)
```

https://github.com/grpc/grpc-kotlin/blob/master/stub/src/main/java/io/grpc/kotlin/ServerCalls.kt#L249-L258
This violates the expected behavior of `onHalfClose`. 
Furthermore, it blocks Armeria's event loop. See https://github.com/grpc/grpc-kotlin/issues/151.

By rescheduling the close event we can avoid the race condition when a close event is called as the result of `onHalfClose`, 

Modifications:
- Close a message deframer when it completes when `HttpStreamReader.closeDeframer` is called.
- Close ArmeriaClientCall when a gRPC client consumes all messages fully.

Result:

- A gRPC-Kotlin sever handles blocking tasks with Coroutine properly
- Fixes #2947

### CODE REVIEW

**Comment by ikhoon** [comment]
The PR diff looks noisy because of the generated code of Kotlin stub. Please take a look at ArmeriaMessageDeframer and HttpStreamReader first. 🙇‍♂️

File: examples/grpc-kotlin/src/main/kotlin/example/armeria/grpc/kotlin/HelloServiceImpl.kt [hunk 1]
```diff
@@ -47,7 +47,22 @@ class HelloServiceImpl : HelloServiceGrpcKt.HelloServiceCoroutineImplBase(Dispat
      */
     override suspend fun blockingHello(request: HelloRequest): HelloReply = withArmeriaBlockingContext {
         try { // Simulate a blocking API call.
-            Thread.sleep(3000)
+            Thread.sleep(10)
```
Comment (trustin) [inline examples/grpc-kotlin/src/main/kotlin/example/armeria/grpc/kotlin/HelloServiceImpl.kt:50]: 3000?

File: examples/grpc-kotlin/src/test/kotlin/example/armeria/grpc/kotlin/HelloServiceTest.kt [hunk 2]
```diff
@@ -52,6 +53,23 @@ class HelloServiceTest {
...
+            val helloService = Clients.newClient(uri, HelloServiceCoroutineStub::class.java)
+            repeat(30) {
+                launch {
+                    val message = helloService.shortBlockingHello(
+                        HelloRequest.newBuilder().setName("$it Armeria").build()
+                    ).message
+                    assertThat(message).isEqualTo("Hello, $it Armeria!")
+                }
+            }
+
```
Comment (trustin) [inline examples/grpc-kotlin/src/test/kotlin/example/armeria/grpc/kotlin/HelloServiceTest.kt:69]: Could remove the empty line

**Comment by ikhoon** [comment]
Ready for review. 😀

File: grpc-protocol/src/main/java/com/linecorp/armeria/common/grpc/protocol/ArmeriaMessageDeframer.java [hunk 3]
```diff
@@ -317,6 +328,12 @@ public void closeWhenComplete() {
      */
     @Override
     public void close() {
+        if (whenClosed == null) {
+            whenClosed = CLOSED_FUTURE;
+        } else {
+            whenClosed.doComplete();
```
Comment (minwoox) [inline grpc-protocol/src/main/java/com/linecorp/armeria/common/grpc/protocol/ArmeriaMessageDeframer.java:334]: Shouldn't we do this after cleanup the resources? which means at the end of this method?

Comment (ikhoon) [reply grpc-protocol/src/main/java/com/linecorp/armeria/common/grpc/protocol/ArmeriaMessageDeframer.java]: Fixed. :-)

**Comment by trustin** [comment]
Thanks, @ikhoon. Scala users will rejoice!

**Comment by ikhoon** [comment]
Thanks for reviewing!

**Pull request merged**